core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -0,0 +1,282 @@
+        // without any bugs in the server implementation. So we ignore this exception and build an error
+        // response for the user.
+        //
+        // For example, the connection could time out before every reaching the server, in which case we'll
+        // likely get a 5xx with the load balancers default 5xx response.
+        LOG.error("Failed to parse an error response. Will create one instead.", e);
+      }
+    }
+
+    if (errorResponse == null) {
+      errorResponse = buildDefaultErrorResponse(response);
+    }
+
+    errorHandler.accept(errorResponse);
+
+    // Throw a RESTException, in case the provided error handler does not throw.
+    throw new RESTException(
+        "Entered the error handler path but no exception was thrown for error response %s",
+        ErrorResponseParser.toJson(errorResponse));
+  }
+
+  /**
+   * Method to execute an HTTP request and process the corresponding response.
+   *
+   * @param method       - HTTP method, such as GET, POST, HEAD, etc.
+   * @param path         - URL path to send the request to
+   * @param requestBody  - Content to place in the request body
+   * @param responseType - Class of the Response type. Needs to have serializer registered with ObjectMapper
+   * @param errorHandler - Error handler delegated for HTTP responses which handles server error responses
+   * @param <T>          - Class type of the response for deserialization. Must be registered with the ObjectMapper.
+   * @return The response entity, parsed and converted to its type T
+   */
+  public <T> T execute(
+      Method method, String path, Object requestBody, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    if (path.startsWith("/")) {
+      throw new RESTException(
+          "Received a malformed path for a REST request: %s. Paths should not start with /", path);
+    }
+
+    String fullUri = String.format("%s/%s", uri, path);
+    HttpUriRequestBase request = new HttpUriRequestBase(method.name(), URI.create(fullUri));
+    addRequestHeaders(request);
+
+    if (requestBody != null) {
+      try {
+        StringEntity stringEntity = new StringEntity(mapper.writeValueAsString(requestBody));
+        request.setEntity(stringEntity);
+      } catch (JsonProcessingException e) {
+        throw new RESTException(e, "Failed to write request body: %s", requestBody);
+      }
+    }
+
+    try (CloseableHttpResponse response = httpClient.execute(request)) {
+
+      // Skip parsing the response stream for any successful request not expecting a response body
+      if (responseType == null && isSuccessful(response)) {
+        return null;
+      }
+
+      String responseBody = responseType != null ? extractResponseBodyAsString(response) : null;
+
+      if (!isSuccessful(response)) {
+        // The provided error handler is expected to throw, but a RESTException is thrown if not.
+        handleFailure(response, responseBody, errorHandler);
```

Comment:
Minor: consider naming the method `throwFailure` rather than `handleFailure` so that it is obvious here that the method is guaranteed to throw an exception.

Reply:
Updated.